README.md
# Wait for Netlify Deployment

Do you have other Github actions (Lighthouse, Cypress, etc) that depend on the Netlify deploy URL? This action will wait until the deploy URL is available before running the next task.

This action uses the Netlify API to always retrieve the correct deployment being built. You will need to generate a [Personal Access Token](https://app.netlify.com/user/applications/personal) to use and pass it as the `NETLIFY_TOKEN` environment variable.

## Env

### `NETLIFY_TOKEN`

**Required.** Your Netlify [Personal Access Token](https://app.netlify.com/user/applications/personal) to use for API access. This should be set as a GitHub secret, see example.

## Inputs

### `site_id`

**Required** The API ID of your site. See Settings > Site Details > General in the Netlify UI

### `deploy_timeout`

Optional — The amount of time to spend waiting on the Netlify deployment to complete the build phase. Defaults to 300 seconds.

### `readiness_timeout`

Optional — The amount of time to spend waiting on the Netlify deployment to respond with a "ready" status. Defaults to 900 seconds.

### `response_timeout`

Optional — The amount of time to spend waiting on the Netlify deployment to respond with a success HTTP code after reaching "ready" status. Defaults to 60 seconds.

### `context`

Optional — The Netlify deploy context. Can be `branch-deploy`, `production` or `deploy-preview`. Defaults to all of them.

## Outputs

### `url`

The Netlify deploy preview url that was deployed.

### `deploy_id`

The Netlify deployment ID that was deployed.

## Example usage

Basic Usage

```yaml
steps:
  - name: Wait for Netlify Deploy
    uses: probablyup/wait-for-netlify-action@3.2.0
    id: waitForDeployment
    with:
      site_id: 'YOUR_SITE_ID' # See Settings > Site Details > General in the Netlify UI
    env:
      NETLIFY_TOKEN: ${{ secrets.NETLIFY_TOKEN }}

# Then use it in a later step like:
# ${{ steps.waitForDeployment.outputs.url }}
```

<details>
<summary>Complete example with Cypress</summary>
<br />

```yaml
name: Cypress
on: pull_request
jobs:
  integration:
    runs-on: ubuntu-latest

    jobs:
    cypress:
      runs-on: ubuntu-latest
      steps:
        - uses: actions/checkout@v2

        - name: Install modules
          run: npm ci

        - name: Wait for Netlify
          uses: probablyup/wait-for-netlify-action@3.2.0
          id: waitForDeployment
          with:
            site_id: '[your site ID here]'
          env:
            NETLIFY_TOKEN: ${{ secrets.NETLIFY_TOKEN }}

        - name: Run Cypress
          uses: cypress-io/github-action@v2
          with:
            record: true
            config: baseUrl=${{ steps.waitForDeployment.outputs.url }}
          env:
            # pass the Dashboard record key as an environment variable
            CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
            # this is automatically set by GitHub
            GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

</details>

<details>
<summary>Complete example with Lighthouse</summary>
<br />

```yaml
name: Lighthouse

on: pull_request

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v1
      - name: Use Node.js 12.x
        uses: actions/setup-node@v1
        with:
          node-version: 12.x
      - name: Install
        run: |
          npm ci
      - name: Build
        run: |
          npm run build
      - name: Waiting for 200 from Netlify
        uses: probablyup/wait-for-netlify-action@3.2.0
        id: waitForNetlifyDeploy
        with:
          site_id: 'YOUR_SITE_ID' # See Settings > Site Details > General in the Netlify UI
        env:
          NETLIFY_TOKEN: ${{ secrets.NETLIFY_TOKEN }}
      - name: Lighthouse CI
        run: |
          npm install -g @lhci/cli@0.3.x
          lhci autorun --upload.target=temporary-public-storage --collect.url=${{ steps.waitForNetlifyDeploy.outputs.url }} || echo "LHCI failed!"
        env:
          LHCI_GITHUB_APP_TOKEN: ${{ secrets.LHCI_GITHUB_APP_TOKEN }}
```

</details>

## Versioning

This project uses [Calver](https://calver.org) to version the apps and builds. The version format is: `YY.0M.MICRO`.

Examples:
  - `22.01.1` - First released build in January of 2022
  - `22.02.15` - Fifteenth release in February of 2022
  - `22.03-beta` - Beta release of the March 2022 build
  - `22.03.4-beta` - Beta release of the fourth build in March 2022

Where `Version` is the two-digit year, the zero-padded two-digit month, and a micro segment to indicate the build number for that month. An optional `-MODIFIER` can be used for pre-release, alpha, beta, and the like.

<small>
This is a heavily-modified fork of [kamranayub/wait-for-netlify-action](https://github.com/kamranayub/wait-for-netlify-action).
</small>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Calver](https://calver.org) with a version format of `YY.0M.MICRO`.

## [Unreleased]

### Added

- Fine-grain controls over timeouts from [@hharnisc](https://github.com/hharnisc)
- Option to filter commits by deploy context [@dnsv](https://github.com/dnsv)
- Docker Compose configuration for development
- `eslint` and configuration
- `prettier` and configuration
- `@vercel/ncc` to build the action as recommended in the GitHub documentation
- CircleCI for running linters and tests
- Dependabot for dependency updates
- GitHub settings to manage repository

### Changed

- Update action to use `node16`
- Update package manager from `npm` to `yarn`
- Update dependencies
- Remove `node_modules` from repository

---

Forked to `mutations/wait-for-netlify-action` - 2022-02-08

- Last version before fork: 3.2.0 [0855dad](https://github.com/mutations/wait-for-netlify-action/commit/0855dade13fa1445d67907d2b5b2ce470e38c468)

---

Changelog sections:

- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Fixed` for any bug fixes.
- `Removed` for now removed features.
- `Security` in case of vulnerabilities.


src/run/run.js
import waitForDeployCreation from '../wait-for-deploy-creation'
import waitForReadiness from '../wait-for-readiness'
import waitForUrl from '../wait-for-url'
import * as core from '@actions/core'
import * as github from '@actions/github'

const STATE_SKIPPED = 'skipped'

const run = async () => {
  try {
    const netlifyToken = process.env.NETLIFY_TOKEN
    const commitSha =
      github.context.eventName === 'pull_request'
        ? github.context.payload.pull_request.head.sha
        : github.context.sha

    const DEPLOY_TIMEOUT = Number(core.getInput('deploy_timeout')) || 60 * 5
    const READINESS_TIMEOUT =
      Number(core.getInput('readiness_timeout')) || 60 * 15
    // keep max_timeout for backwards compatibility
    const RESPONSE_TIMEOUT =
      Number(core.getInput('response_timeout')) ||
      Number(core.getInput('max_timeout')) ||
      60
    const siteId = core.getInput('site_id')
    const context = core.getInput('context')

    if (!netlifyToken) {
      core.setFailed(
        'Please set NETLIFY_TOKEN env variable to your Netlify Personal Access Token secret',
      )
    }
    if (!commitSha) {
      core.setFailed('Could not determine GitHub commit')
    }
    if (!siteId) {
      core.setFailed('Required field `site_id` was not provided')
    }

    let message = `Waiting for Netlify to create a deployment for git SHA ${commitSha}`

    if (context) {
      message += ` and context ${context}`
    }

    console.log(message)
    const commitDeployment = await waitForDeployCreation(
      `https://api.netlify.com/api/v1/sites/${siteId}/deploys`,
      commitSha,
      DEPLOY_TIMEOUT,
      context,
    )

    const url = `https://${commitDeployment.id}--${commitDeployment.name}.netlify.app`

    core.setOutput('deploy_id', commitDeployment.id)
    core.setOutput('url', url)

    console.log(
      `Waiting for Netlify deployment ${commitDeployment.id} in site ${commitDeployment.name} to be ready`,
    )
    const state = await waitForReadiness(
      `https://api.netlify.com/api/v1/sites/${siteId}/deploys/${commitDeployment.id}`,
      READINESS_TIMEOUT,
    )

    if (state === STATE_SKIPPED) {
      core.setOutput(STATE_SKIPPED, 'true')
      console.log('Netlify build was skipped')
    } else {
      core.setOutput(STATE_SKIPPED, 'false')
      console.log(`Waiting for a 200 from: ${url}`)
      await waitForUrl(url, RESPONSE_TIMEOUT)
    }
  } catch (error) {
    core.setFailed(typeof error === 'string' ? error : error.message)
  }
}

export default run
